src/execution_plan/ops/op_apply.c

```diff
@@ -7,97 +7,91 @@
 #include "op_apply.h"
 
 /* Forward declarations. */
+static OpResult ApplyInit(OpBase *opBase);
 static Record ApplyConsume(OpBase *opBase);
 static OpResult ApplyReset(OpBase *opBase);
 static void ApplyFree(OpBase *opBase);
 
 OpBase *NewApplyOp(const ExecutionPlan *plan) {
 	Apply *op = rm_malloc(sizeof(Apply));
-	op->init = true;
-	op->rhs_idx = 0;
-	op->lhs_record = NULL;
-	op->rhs_records = array_new(Record, 32);
+	op->r = NULL;
+	op->op_arg = NULL;
+	op->bound_branch = NULL;
+	op->rhs_branch = NULL;
 
 	// Set our Op operations
-	OpBase_Init((OpBase *)op, OPType_APPLY, "Apply", NULL, ApplyConsume, ApplyReset, NULL, NULL,
+	OpBase_Init((OpBase *)op, OPType_APPLY, "Apply", ApplyInit, ApplyConsume, ApplyReset, NULL, NULL,
 				ApplyFree, false, plan);
 
 	return (OpBase *)op;
 }
 
-static Record ApplyConsume(OpBase *opBase) {
-	Apply *op = (Apply *)opBase;
-
-	assert(op->op.childCount == 2);
+static OpResult ApplyInit(OpBase *opBase) {
+	assert(opBase->childCount == 2);
 
-	Record rhs_record;
-
-	if(op->init) {
-		// On the first call to ApplyConsume, we'll read the entirety of the
-		// right-hand stream and buffer its outputs.
-		op->init = false;
+	Apply *op = (Apply *)opBase;
+	/* The op's bound branch and optional match branch have already been built as
+	 * the Apply op's first and second child ops, respectively. */
+	op->bound_branch = opBase->children[0];
+	op->rhs_branch = opBase->children[1];
```

Comment:
I can't think of any scenario that will violate this assumption currently. For now the direct child of the RHS branch is guaranteed to be an Optional op, so we could explicitly check for that, but I'd prefer to keep Apply's logic more generic so that it can be used for other purposes in the future.